### ISSUE TITLE
Provide subshell like functionality

### ISSUE THREAD

**Issue by jorbsd** [fish-shell/fish-shell#1439]
While fish avoids forking/execing, having subshell like functionality would still be very nice.

In bash/zsh:

pwd; (cd <otherdir>; <dostuff>; pwd;); pwd

should be in the first dir, change and do stuff, then show that you are back in the original did

This closure of shell state in the subshell (vars, directories, etc.) is very handy.

bash/zsh modern syntax uses $() to be similar to () in fish, both replacing ``

**Comment by c22** [comment]
If I'm not mistaken, this problem kinda goes away if #563 is fixed, right?

**Comment by ridiculousfish** [comment]
Maybe, maybe not, depending on how it's fixed. A subshell is both concurrent execution and independent state (multiple $PWDs, etc).

**Comment by d** [comment]
variables (exported or not) is a case where I'd love to have a subshell:

### Scenario
```fish
begin --subshell
    docker-machine env aws-fast | source
    docker run --rm --volume /ccache ubuntu tar c /ccache
end | begin --subshell
    docker-machine env aws-slow | source
    docker run --rm --volume /ccache tar x
end
```

**Comment by ridiculousfish** [comment]
The `begin --subshell` syntax is pretty clever

**Comment by alphapapa** [comment]
My two cents: I'd love to have a `withd`-style command.  My situation: running a shell in Emacs, I want to run `make` in the parent directory, but end up back in the current directory, without having to explicitly save and restore the directory.  This is easy in Bash with subshells, but not in Fish.  A `withd` could be put together as a script now, one that takes a string, but it would preclude syntax highlighting.

**Comment by nrnrnr** [comment]
Either `cd -l` or `withd` would work for me.

**Comment by stephane-chazelas** [comment]
subshells are also useful to isolate a portion of code with regards to signal delivery (think SIGPIPE for instance).

Note that ksh93 _fakes_ subshells without forking (so holds a copy of all original resources (umask, cwd, traps, aliases...) which are restored when the subshell exits). You may want to have a look at that for your `cd -l`, alias -l, umask -l, function -l (like for cd -l open the current directory on a fd with O_CLOSEEXEC, and fchdir to that latter).

**Comment by ridiculousfish** [comment]
That's kind of amazing that ksh93 does that. Thanks for the tip, I definitely want to take a look.

**Comment by aaronjwood** [comment]
If fork/exec was implemented, at least optionally, parallelism could also be restored. For example:
```
sleep 10 &
sleep 10 &
sleep 10 &
sleep 10 &
wait
```
`wait` doesn't block in fish which I think is a real killer. Is there a fish-specific way of doing this, similar to how you have an emulated way of doing subshells?

I'm guessing everything runs in some sort of event loop in fish due to `wait` returning immediately...?

**Comment by alphapapa** [comment]
@aaronjwood Unfortunately there is no `wait` in fish:

```
$ type wait
type: Could not find “wait”
```

**Comment by aaronjwood** [comment]
@alphapapa 
```
type wait
wait is /usr/bin/wait
```
It seems that fish picks up on shell builtins. If not I would think tons of stuff wouldn't work like `cd` or `true`.

**Comment by aaronjwood** [comment]
Actually, my understanding is that `wait` should be a builtin. Similar to:
```
type true
true is a builtin
```
I don't know why fish is picking `wait` up as not a builtin...

Without fish:
```
which true
/usr/bin/true

which wait
/usr/bin/wait

type true
true is a shell builtin

type wait
wait is a shell builtin
```

With fish:
```
which true
/usr/bin/true

which wait
/usr/bin/wait

type true
true is a builtin

type wait
wait is /usr/bin/wait
```

**Comment by alphapapa** [comment]
@aaronjwood I'm not sure what you mean.  Bash provides `wait` as a builtin, but Fish doesn't.  I don't have `/usr/bin/wait` on my systems, either Ubuntu or Debian.  What system are you using?

**Comment by aaronjwood** [comment]
I was on OSX when I did that. Maybe I misunderstood what fish is supposed to be. I thought it's still bash with some nice enhancements on top. Is fish supposed to be a totally custom shell with its own syntax and features?

My understanding was that it's supposed to still be bash in the end. If the goal of fish is not to enhance bash or be compatible with it then I guess you can ignore what I said.

**Comment by krader1961** [comment]
@aaronjwood, I'm using macOS and have no idea how */usr/bin/wait* could work. It does appear to try to do what is typically associated with this command as shown by this error message:

```
/usr/bin/wait: line 4: wait: pid 3906 is not a child of this shell
```

That was from trying to wait for a job I did launch from fish. I don't see that could possibly work as an external command. It is simply a wrapper around the `wait()` or `wait4()` syscall and such can't wait for a process launched by its parent shell. It's like `cd` in that it has to be a builtin of the shell.

> My understanding was that it's supposed to still be bash in the end.

Your understanding is incorrect. Fish deliberately breaks from the POSIX 1003.1 standard that bash is based on. This is something we could possibly document better. Having said that we definitely should have a `wait` builtin. It's just that no one has been sufficiently motivated to implement it. Patches are welcome :smile:

**Comment by aaronjwood** [comment]
Sure, makes sense. Thanks for clarifying about the standard!

**Comment by Debilski** [comment]
By the way: the macOS/FreeBSD `/usr/bin/wait` is no more than a shell script calling the builtin in a slightly convoluted way:

    #!/bin/sh
    builtin `echo ${0##*/} | tr \[:upper:] \[:lower:]` ${1+"$@"}

**Comment by krader1961** [comment]
@Debilski, Can you explain how that works? Because an external command cannot interact with its parent shell in the way this one appears to be doing. This appears to rely on the parent shell sourcing that script for the sole purpose of allowing the user to write `WAIT %jobspec`. That is, converting the `wait` command to all lower case. Which itself doesn't seem to be useful since the user has presumably already typed `wait` in order to find and source that script.

Googling did not result in any useful documents. I did find some, such as [this one](https://www3.physnet.uni-hamburg.de/physnet/Tru64-Unix/HTML/MAN/MAN1/0522____.HTM) which are totally unhelpful in explaining why an external script named `wait` can be used to wait for jobs launched by the parent shell to terminate. This script/command appears to depend on some quirk of BSD that allows a program to wait for an arbitrary process to exit -- not just child processes of the current process.

**Comment by Debilski** [comment]
@krader1961 The lowercasing would be because of case-insensitive filesystems. `WAIT` (or `CD` and a few more commands) will find the correct script there (and then do nothing at all because it obviously doesn’t work).

Here is a little explanation for the rest https://superuser.com/q/69869 . Maybe it makes sense for one of the other commands that are aliased to the very same script.

**Comment by aaronjwood** [comment]
@Debilski well I certainly learned something today! Never realized these were there as "dummy" shell scripts to fulfill POSIX requirements. More good info here that points out a few useful cases such as exit status https://unix.stackexchange.com/a/50060/40967

**Comment by pirate** [comment]
I've implemented my own subshells, wait, & signals for fish if anyone wants to check them out: [`await.fish`](https://github.com/pirate/fish-functions/blob/master/await.fish), [`progress_await.fish`](https://github.com/pirate/fish-functions/blob/master/progress_await.fish), [`signal.fish`](https://github.com/pirate/fish-functions/blob/master/signal.fish), [`background.fish`](https://github.com/pirate/fish-functions/blob/master/background.fish), [`fork`](https://gist.github.com/pirate/01dac89a8db3d79e970365a1f031732f).

Not as ideal as builtin solutions, but it maybe be useful for people looking for interim solutions until backgrounding and subshells are supported natively.

**Comment by arcanis** [comment]
Subshells are useful when one wish to group a command output:

```
(tput a; printf something; tput b) | cat -e
```

**Comment by faho** [comment]
@arcanis: That particular case is actually already possible with `begin` and `end` - `begin; echo banana; echo grapefruit; end | grep banana`.

The issue only starts happening when you're expecting certain state to not be carried over, like $PWD.

**Comment by jtgans** [comment]
The problem with uising `begin; ...; end |grep` is the buffering involved. Unfortunately, this means that if I need to construct streams for large amounts of data, the `begin` and `end` solution doesn't actually solve the problem.

**Comment by aario** [comment]
Simply using back tics ` ` ` and having a subshell and then passing the result to the command like running a loop over results of find command:
```
for file in `find ./ -type f`; ...
```
Is really awesome. I think the only way to have this feature is to fork fish!

**Comment by faho** [comment]
@aario:

>running a loop over results of find command is really awesome

It is!

>I think the only way to have this feature is to fork fish!

Fortunately, it's not:

```fish
for file in (find ./ -type f)
```

Syntactically, the only difference to POSIX-like shells is that you use `()`. Semantically, this will split `find`s output on newlines only instead of spaces, tabs and newlines (or rather the contents of `$IFS`). That's most likely to be a massive improvement because it won't break on files with spaces in the name (which are quite common) - it'll still break with newlines in the filename, but those are quite uncommon.

This issue is about something slightly different - simply put, it's about allowing some form of block that has a separated environment, including the working directory. What you are talking about is called a "command substitution".

----

Also, even in POSIX-like shells like bash, zsh, ksh etc, I'd recommend using `$()` instead of backticks. See http://mywiki.wooledge.org/BashFAQ/082 for more information.

**Comment by aario** [comment]
Didn't ( and ) exist in fish. Gonna give it a try. Thanks.

**Comment by PoolloverNathan** [comment]
All I need is another process so I can put `kill $fish_pid` in my `terms--notification`'s `--action` and stop the loop.

**Comment by krobelus** [comment]
I'd prefer some nice quoting syntax, that works just like other quotes except fish provides syntax highlighting and completions just like with `()`.

Then you would always explicitly specify the executable:

```fish
$ $SHELL -c (🐟 echo I am a fishy subshell 🐟)
$ $SHELL -c <><echo subshell><>
$ $SHELL -c ((echo subshell))
$ $SHELL -c `echo subshell`
```

Right now you can define a simple wrapper command that will give you completions and most syntax highlighting, but you need to do quoting/escaping yourself. I use the backtick (`) here:

```fish
function `
    string join -- " " $argv
end
complete -c ` -xa '(set -l toks (commandline -opc)[2..-1] (commandline -ct); complete -C "$toks")'

# Use it like: fish -c (` some command)
# completion should work after the `
fish -c (` echo Subshell parent: $fish_pid child: \$fish_pid )
```

**Comment by bqv** [comment]
I know this is a 6 year old issue (wow), but could we at least have a `withd` implemented, or is there a sensible alternative now?

**Comment by zanchey** [comment]
@bqv, as we both know, it's a matter of someone doing it - even writing a more concrete spec for the interface would be useful.

**Comment by KingDuckZ** [comment]
> I think the question is whether this is useful for anything other than `cd dir; …`. You might get similar behaviour with `pushd`, `popd` and `set -l` in `begin`…`end` blocks _and_ still retain the chance to escape the inner scope.

As an example, I'm trying to run this bash command that I found on some tutorial online: `(umask 077 && tar -caf $HOME/gnupg-backup_`date +%Y%m%d_%H%M%S`.tar.xz -C $HOME .gnupg)`

**Comment by bqv** [comment]
There's a PR. Now you can all at least properly work out how you want this done

**Comment by bqv** [comment]
PR denied. This should probably just be closed if nobody is going to even otherwise attempt, and any trival implementation will be rejected

**Comment by pirate** [comment]
Unfortunately at our company we're no longer allowed to use fish for writing scripts anymore due to the lack of subshells and advanced signal handling. I still love it as my day-to-day terminal, but I've resorted to bash for writing anything longer than a few lines because of missing subshells.

**Comment by ridiculousfish** [comment]
All, I'm working on it but don't know when it will be done.

**Comment by it-is-wednesday** [comment]
For those of us that are here for the `(cd xxx && ...)` pattern, you can instead use `env --chdir` (with coreutil's `env`).

**Comment by bqv** [comment]
I honestly just started using execline more instead

**Comment by marco-m** [comment]
For people arriving here from an Internet search for "fish shell subshell", there is some documentation on the web site, explaining various approaches with more or less fidelity to a real subshell: https://fishshell.com/docs/current/fish_for_bash_users.html#subshells

**Comment by zwhitchcox** [comment]
I'm trying to use the `set -l` feature, and it is not working for me:

```
# list all directories in current directories
 for p in $(ls); begin; set -l; cd $p; pwd; end; end
```

I'm aware I could use `find . -depth 1 -type d  -exec pwd \;`, but that does not lend itself to longer scripts that could be happening in that directory.

I guess you could use `pushd` and `popd` though.

**Comment by IsaacOscar** [comment]
> I think the question is whether this is useful for anything other than `cd dir; …`. You might get similar behaviour with `pushd`, `popd` and `set -l` in `begin`…`end` blocks _and_ still retain the chance to escape the inner scope.

Subshells could be a way to solve https://github.com/fish-shell/fish-shell/issues/1396

For example,
currently doing something like this:
```
set tr (which tr)
begin
    sleep 2
    echo Hello
end |  begin
   echo '$tarting...'
   $tr l \'
end
```
Will hang for 2 seconds, then print:
```
$tarting...
He''o
```

Whereas
```
begin
    sleep 2
    echo Hello
end |  fish -c "
    echo \\\$tarting...
    "(string escape $tr)" l \\'"
```
Will print `Starting...` imediately, then after 2 seconds, print `He''o`
Notice all the extra escaping I had to do in the `fish-c` version, in particular, I need to use `string escape` in case `$tr` contains special characters.

If we had a `subshell` builtin, we could simply write:
```
begin
    sleep 2
    echo Hello
end |  subshell
   echo '$tarting...'
   $tr l \'
end
```
Instead of the `fish -c`, and so we don't have to worry about escaping the argument, and it would be clear to users that its a "subshell" and wont share variables, workindg directory, etc.

**Comment by kamalmarhubi** [comment]
Would maintainers consider accepting `cd -l` in analogy with `set -l` as a way to unblock the common use case of runnign a command in a different directory? The workarounds listed so far have some annoying shortcomings:
- plain `begin; end` lets the `cd` escape
- `fish -c` doesn't let you use variables defined with `set -l` in the outer shell; they have to be exported
- `env --chdir` has the same issue with unexported local variables, and additionally doesn't let you use fish builtins in the command you run; it also doesn't work on systems that don't use GNU coreutils

A `cd -l` which changes the working directory until the enclosing `end` would have none of those shortcomings. But I can't tell how well or poorly it fits the direction the maintainers have in mind.

**Comment by zanchey** [comment]
Likely blocked on the Rust port, which will make functionality like this a lot more achievable (we think)

**Comment by nrnrnr** [comment]
@alurm  Tcl is not a bad model, but Lua, which is a slightly greater design, allows much more programmer control over nested quotes.  Details at http://lua-users.org/wiki/StringsTutorial, but in brief, the Lua design provides an easy, elegant solution for the problem of "I want to include the end-of-quote character in my quoted string."

**Comment by faho** [comment]
>Perhaps that would be a good idea for Fish, perhaps not.

Terribly sorry, I do not believe it would.

It is sufficiently alien that I don't think it is a good fit for fish.

It also is not really a solution for the subshell idea, because that has the subshell start with the exact state of the outer shell - all variables (not just exported), all functions (not just autoloaded).

**Comment by alurm** [comment]
@nrnrnr,

I'm not so sure about that.

Their comment syntax is interesting indeed though. It has different tradeoffs. For example, `[[]]` is not valid in Lua, but `{{}}` is valid in Tcl.

Perhaps there's a potential to use off-side rule for quoting here as well. I like it because it seems very "correct". I'm not sure it's practical though.

@faho,

I agree, it is very alien. Most languages don't have nesting quoting. I'm not sure if it is a good fit for Fish either.

A possible counter-point could be that Fish is already alien: "end", "psub", "posix-me-harder", "()", and so on. But I don't think it's a very good counter-point, to be honest :)

So, yeah.

> Not really a solution.

Yeah. It's not a 100% solution to a subshell problem.

As I understand, this allows embedding scripts in any language, which uses the same quoting syntax.

Just an example: if `{}` syntax is used, then, as I understand, it will be possible to embed scripts in Tcl easily.

And since Fish uses the same quoting as Fish (obviously), this allows embedding Fish.

To summarize, this is not a potential solution to a subshell problem, but a more abstract (and therefore less useful in particular) problem.

**Comment by mralusw** [comment]
I've just realized that fish *can* create subshells, via backgrounding:
```fish
sleep 1 & wait; echo 1; sleep 2 & wait; echo 2
```

Or, more complex:
```fish
begin; echo b; sleep 1; echo a; end & wait; echo 1\n; \
begin; echo b; sleep 2; echo a; end & wait; echo 2\n
```
> Likely blocked on the Rust port

I thought that was going to be a parallel development instead of replacing the "current" version?

**Comment by akhilman** [comment]
Another subshell use case that works in bash/zsh called `process substitution`. It allows to use a subshell as an argument for `tee` command.
```bash
tar -cz ./ | tee >(md5sum >&2) >(curl -T - http://webdav.server/path/archive.tar.gz) > /dev/null
```

**Comment by ZuBB** [comment]
In the online manual, there is a [section](https://fishshell.com/docs/current/fish_for_bash_users.html#subshells) about subject functionality. But it did not work for me. Also, this issue opened. Do we have outdated/wrong docs, or did I misread something?

**Comment by tmccombs** [comment]
That section says that fish doesn't have subshells, and lists some alternatives that work in _some_ cases.
